File: src/providers/supabase/rules/scope-queries-by-tenant-column.ts
/**
 * supabase-scope-queries-by-tenant-column (correctness)
 *
 * A `.from(table).select(...)` query that selects a tenant/ownership-style
 * column (e.g. `session_id`, `user_id`) but never filters by it (`.eq()`,
 * `.match()`, `.filter()`) reads every row across every tenant instead of
 * scoping to the caller's own data.
 *
 * Detection walks each query chain bottom-up (`CallExpression:exit`, so
 * inner calls are visited before the calls built on top of them) and
 * threads a shared chain-state object outward through `chainObjectCall`,
 * so a `.eq()` two links up the chain is still recorded against the
 * `.select()` that introduced the tenant column.
 */
import { chainObjectCall, isTenantColumnName, memberPropName, parseSelectColumns } from '../utils.js';

interface ChainState {
  selectNode: any;
  tenantColumns: string[];
  filteredColumns: Set<string>;
}

const rule = {
  meta: {
    type: 'suggestion',
    docs: {
      description: 'Supabase queries that select a tenant column must filter by it',
      category: 'correctness',
      rationale:
        'A column like session_id or user_id existing in the schema (and being selected) signals intent to scope rows to one caller, but selecting it is not the same as filtering by it. Without an .eq()/.match()/.filter() on that column, the query returns every row for every tenant, turning a per-user feed into a single shared, cross-user one.',
      docsUrl: 'https://supabase.com/docs/reference/javascript/eq',
      recommended: true,
    },
    messages: {
      missingTenantFilter:
        'This query selects "{{column}}" but never filters by it. Add .eq("{{column}}", ...) (or .match()/.filter()) to scope results to the caller.',
    },
    schema: [],
  },
  create(context: any) {
    const chainStates = new Map<any, ChainState>();
    const selectStates: ChainState[] = [];

    function recordFilteredColumn(state: ChainState, name: unknown) {
      if (typeof name === 'string') state.filteredColumns.add(name);
      else state.filteredColumns.add('*');
    }

    return {
      'CallExpression:exit'(node: any) {
        const prop = memberPropName(node);
        if (!prop) return;

        const objCall = chainObjectCall(node);

        if (prop === 'select' && objCall && memberPropName(objCall) === 'from') {
          const columns = parseSelectColumns(node.arguments?.[0]);
          const tenantColumns = columns.filter(isTenantColumnName);
          if (tenantColumns.length === 0) return;
          const state: ChainState = { selectNode: node, tenantColumns, filteredColumns: new Set() };
          chainStates.set(node, state);
          selectStates.push(state);
          return;
        }

        const state = objCall ? chainStates.get(objCall) : undefined;
        if (!state) return;

        if (prop === 'eq' || prop === 'filter') {
          const colArg = node.arguments?.[0];
          recordFilteredColumn(state, colArg?.type === 'Literal' ? colArg.value : undefined);
        } else if (prop === 'match') {
          const objArg = node.arguments?.[0];
          if (objArg?.type === 'ObjectExpression') {
            for (const p of objArg.properties ?? []) {
              if (p?.type !== 'Property') continue;
              const name =
                p.key?.type === 'Identifier'
                  ? p.key.name
                  : p.key?.type === 'Literal'
                    ? p.key.value
                    : undefined;
              recordFilteredColumn(state, name);
            }
          } else {
            recordFilteredColumn(state, undefined);
          }
        }

        chainStates.set(node, state);
      },

      'Program:exit'() {
        for (const state of selectStates) {
          if (state.filteredColumns.has('*')) continue;
          const missing = state.tenantColumns.find((c) => !state.filteredColumns.has(c));
          if (missing) {
            context.report({
              node: state.selectNode,
              messageId: 'missingTenantFilter',
              data: { column: missing },
            });
          }
        }
      },
    };
  },
};

export const supabaseScopeQueriesByTenantColumnRule = rule;
export default rule;

File: src/providers/supabase/rules/python/scope_queries_by_tenant_column.py
"""supabase-scope-queries-by-tenant-column

Parity with the JS rule: a `.table(x).select(...)` query that selects a
tenant/ownership-style column (e.g. `session_id`, `user_id`) but never
filters by it (`.eq()`, `.match()`, `.filter()`) before `.execute()`.

Walks each query chain bottom-up (`iter_calls_post_order`, the Python
analogue of ESLint's `CallExpression:exit`) so inner calls are visited
before the calls built on top of them, threading chain state outward
through `chain_object_call`. The trailing `.execute()` call carries no
special handling — it just isn't `select`/`eq`/`match`/`filter`, so it's a
no-op link that state passes through untouched.
"""
from __future__ import annotations

import ast
import sys
from pathlib import Path

_PROVIDER = Path(__file__).resolve().parents[2]
if str(_PROVIDER) not in sys.path:
    sys.path.insert(0, str(_PROVIDER))
from utils import (
    call_report_loc,
    chain_object_call,
    is_supabase_table_call,
    is_tenant_column_name,
    iter_calls_post_order,
    member_prop_name,
    parse_select_columns,
)

RULE_KEY = "supabase-scope-queries-by-tenant-column"


def check(tree: ast.AST, path: str, source: str) -> list[dict]:
    out: list[dict] = []
    # node id -> {"select_node": Call, "tenant_columns": [...], "filtered_columns": set}
    chain_states: dict[int, dict] = {}
    select_states: list[dict] = []

    def record_filtered_column(state: dict, name) -> None:
        state["filtered_columns"].add(name if isinstance(name, str) else "*")

    for node in iter_calls_post_order(tree):
        prop = member_prop_name(node)
        if not prop:
            continue

        obj_call = chain_object_call(node)

        if prop == "select" and obj_call is not None and is_supabase_table_call(obj_call):
            columns = parse_select_columns(node.args[0] if node.args else None)
            tenant_columns = [c for c in columns if is_tenant_column_name(c)]
            if not tenant_columns:
                continue
            state = {"select_node": node, "tenant_columns": tenant_columns, "filtered_columns": set()}
            chain_states[id(node)] = state
            select_states.append(state)
            continue

        state = chain_states.get(id(obj_call)) if obj_call is not None else None
        if state is None:
            continue

        if prop in ("eq", "filter"):
            col_arg = node.args[0] if node.args else None
            record_filtered_column(state, col_arg.value if isinstance(col_arg, ast.Constant) else None)
        elif prop == "match":
            obj_arg = node.args[0] if node.args else None
            if isinstance(obj_arg, ast.Dict):
                for k in obj_arg.keys:
                    name = k.value if isinstance(k, ast.Constant) else None
                    record_filtered_column(state, name)
            elif len(node.keywords) > 0:
                for kw in node.keywords:
                    record_filtered_column(state, kw.arg)
            else:
                record_filtered_column(state, None)

        chain_states[id(node)] = state

    for state in select_states:
        if "*" in state["filtered_columns"]:
            continue
        missing = next((c for c in state["tenant_columns"] if c not in state["filtered_columns"]), None)
        if missing:
            node = state["select_node"]
            line, col = call_report_loc(node)
            out.append({"file": path, "line": line, "column": col, "ruleKey": RULE_KEY})

    return out
